store/tikv/coprocessor.go

```diff
@@ -142,19 +178,110 @@ type copIterator struct {
 	store *tikvStore
 	req   *kv.Request
 	tasks []*copTask
-	index int // index indicates the next task to be executed
+
+	mu          sync.RWMutex
+	respGot     int
+	concurrency int
+	respChan    chan *coprocessor.Response
+	errChan     chan error
+	finished    bool
 }
 
-func (it *copIterator) Next() (io.ReadCloser, error) {
-	// TODO: Support `Concurrent` instruction in `kv.Request`.
+// Pick the next new copTask and send request to tikv-server.
+func (it *copIterator) work() {
+	for {
+		it.mu.Lock()
+		if it.finished {
+			it.mu.Lock()
+			break
+		}
+		// Find the next task to send.
+		var task *copTask
+		for _, t := range it.tasks {
+			if t.status == taskNew {
+				task = t
+				break
+			}
+		}
+		if task == nil {
+			it.mu.Unlock()
+			break
+		}
+		task.status = taskRunning
+		it.mu.Unlock()
+		resp, err := it.handleTask(task)
+		if err != nil {
+			it.errChan <- err
+			break
+		}
+		if !it.req.KeepOrder {
+			it.respChan <- resp
+		} else {
+			task.respChan <- resp
+		}
+	}
+}
 
-	var backoffErr error
-	for backoff := rpcBackoff(); backoffErr == nil; backoffErr = backoff() {
-		if it.index < 0 || it.index >= len(it.tasks) {
+func (it *copIterator) run() {
+	// Start it.concurrency number of workers to handle cop requests.
+	for i := 0; i < it.concurrency; i++ {
+		go it.work()
+	}
+}
+
+// Return next coprocessor result.
+func (it *copIterator) Next() (io.ReadCloser, error) {
+	if it.finished {
+		return nil, nil
+	}
+	var (
+		resp *coprocessor.Response
+		err  error
+	)
+	// If data order matters, repsone should be returned in the same order as copTask slice.
+	// Otherwise all responses are returned from a single channel.
+	if !it.req.KeepOrder {
+		// Get next fetched resp from chan
+		select {
+		case resp = <-it.respChan:
+		case err = <-it.errChan:
+		}
+	} else {
+		var task *copTask
+		it.mu.Lock()
+		for _, t := range it.tasks {
+			if t.status != taskDone {
+				task = t
+				break
+			}
+		}
+		it.mu.Unlock()
+		if task == nil {
+			it.Close()
 			return nil, nil
 		}
-		task := it.tasks[it.index]
+		select {
+		case resp = <-task.respChan:
+		case err = <-it.errChan:
+		}
+	}
+	if err != nil {
+		it.Close()
+		return nil, err
+	}
+	it.mu.Lock()
+	defer it.mu.Unlock()
+	it.respGot++
+	if it.respGot == len(it.tasks) {
+		it.Close()
+	}
```

Comment:
Is `respGot` redundant? Even without it, the iterator will be closed after all tasks' status are _Done_.